File: easy/110_Balanced_Binary_Tree.py
# Given a binary tree, determine if it is height-balanced.
#
# For this problem, a height-balanced binary tree is defined as:
#
# a binary tree in which the left and right subtrees of every node differ in height by no more than 1.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        if not root:
            return 1

        left = self.isBalanced(root.left)
        right = self.isBalanced(root.right)

        if left == 0 or right == 0 or abs(left - right) > 1:
            return False

        return max(left, right) + 1

File: go/easy/110_Balanced_Binary_Tree.go
package main

//Given a binary tree, determine if it is
//height-balanced
//
//Example 1:
//Input: root = [3,9,20,null,null,15,7]
//Output: true
//
//Example 2:
//Input: root = [1,2,2,3,3,null,null,4,4]
//Output: false
//
//Example 3:
//Input: root = []
//Output: true

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	dif := maxDepth_(root.Left) - maxDepth_(root.Right)
	if dif <= 1 && dif >= -1 {
		return isBalanced(root.Left) && isBalanced(root.Right)
	}
	return false
}

func maxDepth_(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return 1 + max_(maxDepth_(root.Left), maxDepth_(root.Right))
}

func max_(x int, y int) int {
	if x > y {
		return x
	}
	return y
}
